test/e2e/logicModules/KPIRewarderLifecycle.t.sol

```diff
@@ -0,0 +1,527 @@
+            }
+        }
+
+        //--------------------------------------------------------------------------------
+        // Test Context Setup
+        //--------------------------------------------------------------------------------
+
+        // Generate Users and Amounts
+        users = new address[](TOTAL_USERS);
+        amounts = new uint[](TOTAL_USERS);
+
+        for (uint i = 0; i < TOTAL_USERS; i++) {
+            users[i] = vm.addr(i + 1);
+            amounts[i] = bound((i * 1000e18), 1, 100_000_000e18);
+        }
+        _assumeValidAddresses(users);
+
+        // Mint enough USDC to the participants
+        _setupUSDC();
+
+        // give the automation service the rights to post assertions
+        _prepareKPIRewarder();
+
+        // Initialize KPIRewarder setup:
+        rewardToken.mint(address(this), REWARD_DEPOSIT_AMOUNT);
+        rewardToken.approve(address(fundingManager), REWARD_DEPOSIT_AMOUNT);
+        fundingManager.deposit(REWARD_DEPOSIT_AMOUNT);
+
+        //--------------------------------------------------------------------------------
+        // Test Rounds
+        //--------------------------------------------------------------------------------
+
+        // Now we loop through the rounds. First round happens with no stakers and the rewards are not distributed, but the reward value is modified. Since the funds stay in the FundingManager until claiming, there are no locked funds in the KPIRewarder.
+
+        uint roundCounter = 0;
+        uint[] memory accumulatedRewards = new uint[](TOTAL_USERS);
+
+        uint totalDistributed = 0;
+        uint totalExpectedRewardsDistributed;
+
+        do {
+            (uint rewardsDistributed, uint expectedDistributed) =
+                _processRound(accumulatedRewards, roundCounter);
+            totalDistributed += rewardsDistributed;
+            totalExpectedRewardsDistributed += expectedDistributed;
+
+            roundCounter++;
+        } while (roundCounter < DEPOSIT_ROUNDS);
+
+        // Withdraw all funds and check balances
+
+        for (uint i; i < TOTAL_USERS; i++) {
+            vm.prank(users[i]);
+            kpiRewarder.unstake(amounts[i]);
+
+            /*console.log(
+                "Current staking Balance:",
+                stakingToken.balanceOf(address(kpiRewarder))
+            );*/
+
+            assertEq(kpiRewarder.balanceOf(users[i]), 0);
+            assertEq(stakingToken.balanceOf(users[i]), amounts[i]);
+            assertEq(rewardToken.balanceOf(users[i]), accumulatedRewards[i]);
+        }
+
+        // Check final invariants
+
+        assertEq(
+            rewardToken.balanceOf(address(fundingManager)),
+            (REWARD_DEPOSIT_AMOUNT - totalDistributed)
+        );
+
+        // TODO: Fix Staking precision bug
+        assertApproxEqAbs(
+            totalDistributed, totalExpectedRewardsDistributed, 1e6
+        );
+
+        /*
+        console.log("Final Rewards Distributed: ", totalDistributed);
```

Comment:
Still logged

Reply:
Do we have a general policy about keeping/removing outcommented logs in tests @marvinkruse ? I personally like keeping them so when getting back to the code in the future I just need to uncomment a couple of lines and can get some extra information about contract state. But I see it can be considered bloat, so I'm happy to remove it if preferred. (Of course I'm referring to well written logs, not just "log("run arrived here") )

Reply:
I just marked it as it wasnt a style we used before
If I wanted to check for a specific value I always just did it with asserts (at least in the e2e tests)
But I can understand if the usage of an external contract is a bit tricky here. But technically all the internal stuff of the staking should ideally be pre conceptualized without any further interactions with the external contracts, no?
I dont have a hard opinion on this I was just wondering

Reply:
Yeah, the cleaner the better, but I'm fine for us to keep outcommented code/logs in there if there is a high chance of us needing it again. Otherwise we should probably remove it. But it's a good reminder to keep these, for when we are revisiting.